geonode/layers/api/serializers.py

```diff
@@ -195,6 +195,37 @@ class Meta:
 
     featureinfo_custom_template = FeatureInfoTemplateField()
 
+    def update(self, instance, validated_data):
+        super().update(instance, validated_data)
+
+        # Handle updates to attribute_set
+        allowed_fields = ["pk", "description", "attribute_label", "visible", "display_order"]
+        if "blob" in validated_data and "attribute_set" in validated_data["blob"]:
```

Comment:
Not sure what you mean, in the example you linked 
`_data = validated_data.pop("blob")`
data is still read from the "blob" field? Do you mean it should be removed after reading like in thie example?

Reply:
@kilichenko-pixida could you debug into the flow and adjust the `AttributeSerializer` instead of handling updates in the `DatasetSerializer` instead? For what I learned when debugging the flow by myself, DRF leverages the functionality/methods provided by the `DynamicRelationField` itself to validate and update fields marked as writable. Place a debug point here and check what actually happens

Reply:
> 
> For what I understand from the code right now, I fear you do not get any `validated_data` [from the `to_internal_value()`](https://github.com/encode/django-rest-framework/blob/6f7aad8ffa96226672eacb82ab003dcf01070ed3/rest_framework/serializers.py#L470) because [the `attribute_set` field is not writable](https://github.com/Thuenen-GeoNode-Development/geonode/blob/8325e6e8dc66db88fe389f58280a535f659e3160/geonode/layers/api/serializers.py#L194).

That was the first thing I changed, but that wasn't a problem and even if writable the validation would fail.

Reply:
> @kilichenko-pixida could you debug into the flow and adjust the `AttributeSerializer` instead of handling updates in the `DatasetSerializer` instead? For what I learned when debugging the flow by myself, DRF leverages the functionality/methods provided by the `DynamicRelationField` itself to validate and update fields marked as writable. Place a debug point here and check what actually happens

I am not sure I am following. For example, if I add an update method to `AttributeSerializer` it wouldn't even be involved when doing this patch. How do you mean to adjust the  `AttributeSerializer`?

Reply:
> That was the first thing I changed, but that wasn't a problem and even if writable the validation would fail.

So, failed for what reason? You can implement `validate_<field>` methods (for what I could spot in the code).

> For example, if I add an update method to AttributeSerializer it wouldn't even be involved when doing this patch

How do you think it is not involved? I am not an expert in this and I did have a detailed look in how this seems to work, but for what I could find out there is [a field validation](https://github.com/encode/django-rest-framework/blob/6f7aad8ffa96226672eacb82ab003dcf01070ed3/rest_framework/serializers.py#L476) going on. The `AttributeSerilizer` in fact [is known to the `attribute_set` field](https://github.com/Thuenen-GeoNode-Development/geonode/blob/8325e6e8dc66db88fe389f58280a535f659e3160/geonode/layers/api/serializers.py#L194). I would suspect that the `DynamicRelationField` provides a mechanism for letting you implement validation. Did you have [a look here already](https://www.django-rest-framework.org/api-guide/validators/)?

Reply:
> So, failed for what reason? You can implement `validate_<field>` methods (for what I could spot in the code).
> 
> How do you think it is not involved? I am not an expert in this and I did have a detailed look in how this seems to work, but for what I could find out there is [a field validation](https://github.com/encode/django-rest-framework/blob/6f7aad8ffa96226672eacb82ab003dcf01070ed3/rest_framework/serializers.py#L476) going on. The `AttributeSerilizer` in fact [is known to the `attribute_set` field](https://github.com/Thuenen-GeoNode-Development/geonode/blob/8325e6e8dc66db88fe389f58280a535f659e3160/geonode/layers/api/serializers.py#L194). I would suspect that the `DynamicRelationField` provides a mechanism for letting you implement validation. Did you have [a look here already](https://www.django-rest-framework.org/api-guide/validators/)?

The full error trace I posted [here](https://github.com/Thuenen-GeoNode-Development/Sprints/issues/14#issuecomment-1822758782)

By not involved I just mean that I tried overriding the update method of the `AttributeSerilizer`, but I remeber it didn't even run when doing this patch.

No, I didn't know I could reimplement `validate_field` methods, I'll look into the docs you just linked and see whether I can indeed write my own validation/serizliation logic for the `DynamicRelationField.`

Reply:
Have you tried looking at https://www.django-rest-framework.org/api-guide/serializers/#writing-update-methods-for-nested-representations ?

Reply:
> I mean, it is not wrong to place things in the `DatasetSerializer`. For what I am concerned about is the `blob` or `data` parts. Just send the (unwrapped) `attribute_set` part.

Yeah, I realize it would be preferrable to not have it wrapped and I don't have any strong opinions on where it should be handled if you think there is a better place for it than `DatasetSerializer`. 

Just sending `attribute_set` unwrapped would fail the default validation/convertion to internal values and not even be passed to the `update` method. Maybe it's as obvious and as simple as overriding some method to avoid default validation but I just lack the background - I'd appreciate a hint if you have one, otherwise I'd dig into documentation after I'm back and try to do it.